File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverse(ListNode* head, int times) {
        ListNode* prev = nullptr;
        ListNode* curr = head;

        while (times--) {
            ListNode* currNext = curr->next;
            curr->next = prev;
            prev = curr;
            curr = currNext;
        }
        return prev;
    }

    ListNode* reverseKGroup(ListNode* head, int k) {

        if (head == nullptr) return head;

        ListNode* left = head;
        ListNode* prevLeft = nullptr;
        ListNode* ans = nullptr;

        while (left) {
            ListNode* right = left;

            // check if k nodes exist
            for (int i = 1; i < k; i++) {
                if (!right->next) {
                    if (prevLeft)
                        prevLeft->next = left;
                    return ans ? ans : head;
                }
                right = right->next;
            }
            ListNode* nextLeft = right->next;

            ListNode* newGrpHead = reverse(left, k);

            if (ans == nullptr)
                ans = newGrpHead;

            if (prevLeft)
                prevLeft->next = newGrpHead;
            prevLeft = left;
            left = nextLeft;
        }

        return ans ? ans : head;
    }
};

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */


var reverse = function(head, k) {
        // code here
        let prev = null;
        let curr = head;
        while(k--){
            let currNext = curr.next;
            curr.next = prev;
            prev = curr;
            curr = currNext;
        }
        return prev;
};
var reverseKGroup = function(head, k) {

    if(head == null)
        return head;
    let prevLeft = null;
    let left = head;
    let ans = null;
    while(left){
        let right = left;
        for(let i=1; i<k; i++){
            if(!right.next){
                if(prevLeft)
                    prevLeft.next = left;
                return ans ? ans : head;
            }
            right = right.next;
        }

        let nextLeft = right.next;
        let newGrpHead = reverse(left, k);
        if(ans == null)
            ans = newGrpHead;
        if(prevLeft)
            prevLeft.next = newGrpHead;
        prevLeft = left;
        left = nextLeft;
    }

    return ans;
};
